Fix digamma integral

rogerbalsach opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #26523
Fixes #26707


#### Brief description of what is fixed or changed

Added regression tests
Implemented the functions _eval_is_finite and _eval_as_leading_term for the lerchphi class.
`lerchphi._eval_is_finite` has been implemented with finite complex numbers in mind. A possible extension to handle infinite arguments is left for the future.
The divergences considered are the following:
$$\Phi(0, s, a) = \frac{1}{a^s}$$
Which diverges when $a=0$ and $s \in \mathbb{R}^+$.
When $z \neq 0$, the function diverges when $a \in - \mathbb{N}$ and $s \in \mathbb{R}^+$.
$$\Phi(z, s, -k) = \sum\_{n=0}^\infty \frac{z^n}{(n-k)^s}$$
Finally, for $z=1$, the function has the same poles as the zeta function: $s = 1$.
No analytic continuation has been considered, except for $\Phi(1,s,a)=\zeta(s,a)$.
`lerchphi._eval_as_leading_term` has been implemented. It handles the following limits:
$$\lim\_{z \to 1} \Phi(z, 1, a_0) = -\log(\log(z)) - \gamma_E - \psi(a_0)$$
If $s \in \mathbb{N}^+$ or $\Re(s_0) > 1$:
$$\lim\_{(z, s, a) \to (1, s_0, a_0)} \Phi(z, s, a) = \zeta(s_0, a_0)$$
If $\Re(s_0) < 1$:
$$\lim\_{(z, s) \to (1, s_0)} \Phi(z, s, a) = \Gamma(1-s) (-\log(z))^{s-1}$$
If $a_0 \in -\mathbb{N}$ and $s \in \mathbb{R}^+$:
$$\lim\_{a \to a_0} \Phi(z, s, a) = \frac{z^{-a_0}}{(a-a_0)^s}$$

Added new functionality to the `zeta._eval_as_leading_term` method.
It adds the following cases:
$$\lim\_{(s,a) \to (s_0, a_0)} \zeta(s, a) = \zeta(s_0, a_0), \qquad \text{if} \qquad s_0 \neq 1$$
$$\lim\_{(s,a) \to (1, a_0)} \zeta(s, a) = \frac{1}{s-1}$$
Added tests for all the new changes.

#### Other comments

Road map:

Following the discussion at https://github.com/sympy/sympy/issues/26523, in order to solve the bug with the integral
$$\int\_0^1 \frac{1-t^z}{1-t} \mathrm{d}t = \psi(z+1) + \gamma_E$$
We need to correctly compute the following limit:
$$\lim\_{t \to 1^-} \left\[-t^{z+1}\Phi(t, 1, z+1) - \log(t - 1)\right\] = \psi(z+1) + \gamma_E - i \pi$$
Sympy currently gives the wrong result for this limit, which causes the bug. To avoid returning the wrong value, we need the following:
We first need sympy to identify that $\Phi(1, 1, z+1)$ is infinite to avoid simply substituting $t=1$. This requires the implementation of `lerchphi._eval_is_finite`.
This is enough to solve the bug in the `gruntz` algorithm. However, Sympy will also try to solve the limit using the `heuristics` algorithm, which also has this bug.
The `heuristics` algorithm seems to just do the substitution $t = 1$. So a check to make sure the result is finite is needed.
With these, the integral should no longer return the wrong result but rather either raise an error or return the unevaluated result.

In order to compute the correct limit, an implementation of `lerchphi._eval_as_leading_term` is needed, which should be able to correctly identify
$$\lim\_{\varepsilon \to 0^+} \Phi(1-\varepsilon, 1, z+1) = -\log(\varepsilon) - \gamma_E - \psi(z+1)$$
Finally, a simplification needs to be added somewhere to simplify
```python
(-t*t**z*z*lerchphi(t, 1, z + 1)*gamma(z + 1)/gamma(z + 2)
 - t*t**z*lerchphi(t, 1, z + 1)*gamma(z + 1)/gamma(z + 2)
 - log(t - 1))
```
to
```python
-t**(z + 1)*lerchphi(t, 1, z + 1) - log(t - 1)
```

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* functions
  * Implemented `lerchphi._eval_is_finite`
  * Implemented `lerchphi._eval_as_leading_term`
  * Added functionality to `zeta._eval_as_leading_term`

* printing
  * Changed Mathematica translation of `lerchphi` to `HurwitzLerchPhi`

<!-- END RELEASE NOTES -->

rogerbalsach committed:
Add regression tests

rogerbalsach committed:
Implement LerchPhi.is_finite

oscarbenjamin commented on sympy/functions/special/zeta_functions.py:
```diff
@@ -214,6 +214,16 @@ def _eval_rewrite_as_zeta(self, z, s, a, **kwargs):
     def _eval_rewrite_as_polylog(self, z, s, a, **kwargs):
         return self._eval_rewrite_helper(polylog)
 
+    def _eval_is_finite(self):
+        z, s, a = self.args
+        if z == 0 or a == 0:
+            return z.is_zero and a.is_zero and s.is_nonpositive
```
What should happen here if any of the `is_` checks gives `None`?

https://docs.sympy.org/latest/guides/booleans.html

rogerbalsach replied:
Hi @oscarbenjamin , thanks for taking a look and the information about the booleans.
I thought python and was compatible with the fuzzy logic but it is indeed not. I will use the fuzzy_and function, this should handle correctly the None cases.
Also, I am not sure what is the convention when the function evaluates to NaN, should the is_finite be None?

oscarbenjamin replied:
> what is the convention when the function evaluates to NaN, should the is_finite be None?

Probably. What case evaluates to nan that you are thinking of?

rogerbalsach replied:
It is because a lot of situations where the function is infinite is because one has $0^{-s}$. The previous check `s.is_nonpositive` gives False when s is not real. I have added a check for s to be real, when s is complex the expression is undefined so now the function should return None.

oscarbenjamin commented on sympy/functions/special/zeta_functions.py:
```diff
@@ -214,6 +214,25 @@ def _eval_rewrite_as_zeta(self, z, s, a, **kwargs):
+            if s.is_real:
+                return s.is_nonpositive
+            return
+        exp_expr = self.expand(func=True)
+        if exp_expr != self:
+            return exp_expr.is_finite
```
What is `expand(func=True)` being used for here?

We should avoid creating new expressions during an assumptions query. Just creating the new expression runs the risk of triggering an identical assumptions query which leads to infinite recursion.

rogerbalsach replied:
Ok, I did not think about that. My idea was that, since LerchPhi includes a lot of common functions as particular values (zeta function, polylog, Dirichlet eta, ...), I can try to write the function in terms of these other functions and then let them determine whether the function is finite or not. Is there a safe way to do it without the risks you are mentioning?

oscarbenjamin replied:
There is not really a safe way to do it. We just have to accept that the core assumptions cannot resolve all queries:
https://docs.sympy.org/latest/guides/assumptions.html

rogerbalsach replied:
Ok, sorry for the delay, but I haven't had time these last few weeks. I understand then that this part of the code should be changed and all the checks need to be implemented from scratch. Is that correct?

oscarbenjamin replied:
Yes, the checks need to be from scratch. It is also acceptable that some assumptions queries just return None.

rogerbalsach commented on sympy/functions/special/zeta_functions.py:
```diff
@@ -216,13 +216,22 @@ def _eval_rewrite_as_polylog(self, z, s, a, **kwargs):
+            if a.is_zero is False:
+                return True
+            if s.is_real:
+                return s.is_nonpositive
```
I'm not sure if this works correctly when a.is_zero is None.

rogerbalsach committed:
Implemented lerchphi._eval_as_leading_term

Implemented functions lerchphi._eval_as_leading_term for a few selected
cases.
Added new cases to the function zeta._eval_as_leading_term needed for
the lerchphi implementation.
Added tests for both implementations.

rogerbalsach committed:
Modified heuristics function and added gammasimp

Modified heuristics function to check for infinite quantities.
Added gammasimp in limit function to simplify the coefficients before
checking for independence on newz.

rogerbalsach committed:
Add author

rogerbalsach commented:
@oscarbenjamin This PR is ready for review. I don't know why the authors tests are failing, can you tell me if I added myself to the author list correctly?

oscarbenjamin commented on sympy/functions/special/zeta_functions.py:
```diff
@@ -214,6 +214,60 @@ def _eval_rewrite_as_zeta(self, z, s, a, **kwargs):
+            return
+        if a.is_integer and a.is_nonpositive:
+            if s.is_real:
+                return s.is_nonpositive
```
Is this correct:
```python
In [21]: z, s, a = 10, -1, -10

In [22]: lerchphi(z, s, a).is_finite
Out[22]: True

In [23]: lerchphi(z, s, a).evalf()
...
ValueError: Lerch transcendent complex infinity
```

oscarbenjamin replied:
Okay, I see what you mean. It does seem as if the function should be continuous there:
```
In [26]: lerchphi(10, -1, 10 + 1e-5).n()
Out[26]: -0.987655432098765

In [27]: lerchphi(10, -1, 10 - 1e-5).n()
Out[27]: -0.987653209876543
```

rogerbalsach replied:
Yes, that's another way. You checked for $a = 10$, which is not problematic, but the same can be done for the problematic point $a = -10$
```
In [26]: lerchphi(10, -1, -10 + 1e-5).n()
Out[26]: 1.23456679012452

In [27]: lerchphi(10, -1, -10 - 1e-5).n()
Out[27]: 1.23456901218742
```
which corresponds to the value given by WA of $\frac{100}{81} = 1.2345679012345679$

oscarbenjamin replied:
Flint's `acb_dirichlet_lerch_phi` gives the same:
```python
In [22]: from flint import acb, ctx

In [23]: ctx.dps = 20

In [24]: acb(10).lerch_phi(-1, -10)
Out[24]: [1.23456790 +/- 9.62e-9] + [+/- 4.91e-9]j
```

rogerbalsach reviewed (commented):
@oscarbenjamin I'm not an expert, but I would say this is a problem with mpmath, if you look at their code they raise this error whenever $a$ is a negative integer (except if $z=0$ or $z=1$). This is because in the lerchphi you have $\frac{1}{(n+a)^s}$, so it diverges when $a = -n$.
If you want this behavior I can simply change lines 229 and 230 to simply "return False"

However, if $s$ is negative that term is no longer divergent.
For example:
[lerchphi(10, -1, -10)](https://www.wolframalpha.com/input?i=HurwitzLerchPhi%5B10%2C-1%2C-10%5D)
[lerchphi(0.5, -1, -10)](https://www.wolframalpha.com/input?i=HurwitzLerchPhi%5B0.5%2C-1%2C-10%5D)

(sorry, I was using the Reply option instead of the Review, so I just now figured out that you didn't receive the answer)

rogerbalsach commented:
@oscarbenjamin, @anutosh491 . Hi, any news regarding this PR? Should I fix the conflict or better wait until it is ready for merging?

anutosh491 commented:
Haven't had a chance to look into this. Will have a look tomorrow !

rogerbalsach commented:
@oscarbenjamin  I've updated and cleaned up the code a bit, and rebased everything on top of the current master.

However, I'm not sure how to address the failing tests. For the author-related test, I added myself to the authors list. The other two failures seem to come from the test that computes `lerchphi(oo, 2, 2)`. I'm not sure whether the best approach is to remove this test or if there is a way to fix the underlying issue.
